Shared termset search

ptbrowne opened the pull request:
The main change of this PR is to change from having a termset centered UI to a dimension centered UI. It should hopefully make the UI more relatable to the users as termsets are not connected to cube but dimensions. To be able to show dimensions with their termsets inside the search results, I switched to a CONSTRUCT query. The storybook is updated, dimensions can be hovered to show the termsets.

- [x] Can search by shared termsets
- [x] Search query uses construct which makes it powerful (can fetch dimensions, termsets, time units for each cube)
- [x] Add dataset dialog shows dimensions and hovering shows termsets
<img width="619" alt="image" src="https://github.com/visualize-admin/visualization-tool/assets/465582/a156cce1-b5c7-4035-a056-d0591ff6d7eb">

- [x] Add dataset dialog has a filter where you can select which dimensions should be considered when searching searchable datasets
- [x] Added stories to showcase join by on multiple cubes
- [x] Temporal dimensions should be shown as "shared dimension" in search results
- [ ] Join by should support multiple dimensions

bprusinowski commented:
One thing I noticed currently on TEST: we can merge the same cube to the currently selected one, which results in no joinBy dimension error. Just thought I would drop it here, maybe it could be connected to this PR? 🤔

ptbrowne commented:
I think Bartosz that you can review this PR as I will tackle the ability to join by several dimensions in another PR. Thanks 🙇 .

bprusinowski commented on app/rdf/parse-search-results.ts:
```diff
@@ -29,6 +31,7 @@ function buildSearchCubes(quads: Quad[]): SearchCube[] {
     if (cubeQuads) {
       const themeQuads = cubeQuads.get(ns.dcat.theme.value);
       const subthemesQuads = cubeQuads.get(ns.schema.about.value);
+      const dimensions = cubeQuads.get("https://visualize.admin.ch/contains");
```
And in the scores query :)

ptbrowne replied:
Done after.

bprusinowski commented on app/components/tag.tsx:
```diff
@@ -1,4 +1,7 @@
 import { BoxProps, Typography, TypographyProps, styled } from "@mui/material";
+import { Theme } from "@mui/material/styles";
+import { makeStyles } from "@mui/styles";
+import cx from "classnames";
```
Nit: same as `clsx` we already use in the application? Is so, maybe we could stick to it? 🤔

bprusinowski commented on app/rdf/parse-search-results.ts:
```diff
@@ -5,6 +5,12 @@ import { SearchCube } from "@/domain/data";
 import { DataCubePublicationStatus } from "@/graphql/resolver-types";
 import * as ns from "@/rdf/namespace";
 
+const visualizePredicates = {
+  hasDimension: "https://visualize.admin.ch/hasDimension",
+  hasTermset: "https://visualize.admin.ch/hasTermset",
+  hasTimeUnit: "https://visualize.admin.ch/hasTimeUnit",
+};
```
💯 Still think it could be useful to extract visualize namespace, I think there were other places there it was used 👀

ptbrowne replied:
We actually already had it yes in rdf/namespace. I reused it now.

bprusinowski commented on app/rdf/parse-search-results.spec.ts:
```diff
@@ -307,3 +316,33 @@ describe("parse-search-results", () => {
+    );
+    return temporal && shared;
+  });
+  expect(cubeWithTemporalAndShared).toBe(true);
```
👌

bprusinowski reviewed (approved):
LGTM, clear code and a super cool feature! 👏 💯 

A few things I noticed when testing the feature using `Photovoltaikanlagen` dataset:
- I could remove dimension filters completely and try to add non-compatible cube (e.g. Forest fire danger), which didn't do anything
- Haven't checked the design, but feels like termsets could also be yellow? 🤔 
<img width="395" alt="Screenshot 2024-04-19 at 16 54 50" src="https://github.com/visualize-admin/visualization-tool/assets/52032047/6b1482e4-5fcf-4314-87a6-0fca29c99f64">

- Looks like the editor filters are not aligned with interactive time range filters (might not be related to this PR), see 2009 in the editor and 2014 in interactive filter 👀 

<img width="1728" alt="Screenshot 2024-04-19 at 17 35 27" src="https://github.com/visualize-admin/visualization-tool/assets/52032047/fbcc4594-ffca-4589-9100-115280acef5d">

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -91,24 +101,53 @@ export const DatasetDialog = ({
     throw new Error("Could not find active chart config");
   }
+
+  const relevantCubes = activeChartConfig.cubes.slice(0, 1);
```
It is to prepare for the future since we are expected to be able to merge several cubes. Step by step 😅 .

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -91,24 +101,53 @@ export const DatasetDialog = ({
     },
   });
 
+  // Getting cube termsets, to then search for cubes with at least one this termset
```
Done.

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -186,27 +269,78 @@ export const DatasetDialog = ({
+            onChange={handleChangeSharedDimensions}
+            input={
+              <OutlinedInput
+                notched={false}
```
I had to pass it otherwise it had one, maybe it's false in more recent versions from MUI.

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -413,6 +412,7 @@ const useAddDataset = () => {
       currentComponents: DataCubeComponents;
       otherCube: PartialSearchCube;
     }) => {
+      currentTermsets;
```
Removed in a later commit.

ptbrowne commented on app/components/data-download.tsx:
```diff
@@ -321,6 +322,7 @@ const DownloadMenuItem = ({
 }) => {
   const locale = useLocale();
   const i18n = useI18n();
+  const client = useClient();
```
It was related to storybook. In storybook, we have to execute against localhost:3000 and not against / (otherwise it tried localhost:6006/graphql). It was working already for useComponentsQuery etc... where the hook execute the query against the provider client, since in Storybook we provide a client correctly configured. But it was not working for functions using directly the imported client. Thus, the change :)

bprusinowski replied:
Thanks for explaining 👌

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -249,7 +249,22 @@ const mkScoresQuery = (
+  ${pragmas}
+  #pragma join.bind off
+
+  PREFIX meta: <https://cube.link/meta/>
+  PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+  PREFIX schema: <http://schema.org/>
+  PREFIX sh: <http://www.w3.org/ns/shacl#>
+  PREFIX cube: <https://cube.link/>
+  PREFIX cubeMeta: <https://cube.link/meta/>
+  PREFIX dcterms: <http://purl.org/dc/terms/>
+  PREFIX dcat: <http://www.w3.org/ns/dcat#>
```
More readable and copy pastable :)

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -416,28 +428,35 @@ export const DatasetDialog = ({
-    left: leftDim.iri,
-    right: rightDim.iri,
-  };
+  console.log({ possibleJoinBys });
```
Removed after, thanks.

ptbrowne commented on app/domain/data.ts:
```diff
@@ -407,7 +407,7 @@ export type SearchCube = {
     iri: string;
     label: string;
-    termsets?: {
+    termsets: {
```
Yes, one, fixed :)

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -411,6 +401,17 @@ export const DatasetDialog = ({
+                        currentComponents,
+                        otherCube: cube,
+                      });
+                      handleClose(ev, "escapeKeyDown");
```
I honestly never know what to put here.

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -169,6 +169,8 @@ const mkScoresQuery = (
       visualize:hasDimension ?dimensionIri.
 
       ?dimensionIri visualize:hasTermset ?termsetIri .
+      ?dimensionIri visualize:hasTimeUnit ?unitType .
```
Since we construct the query, I want to link the dimension with time unit in one hop to facilitate the parsing (`?dimension cubeMeta:dataKind/time:unitType` is two hops).

bprusinowski replied:
Ah I see, misunderstood its purpose 👍

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -416,28 +428,35 @@ export const DatasetDialog = ({
-  }
+  const possibleJoinBys = leftOptions
+    .map((o) => {
+      if (o.type === "temporal") {
```
Done.

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -263,25 +146,31 @@ const mkScoresQuery = (
+    schema:creator ?creatorIri ;
+    schema:workExample <https://ld.admin.ch/application/visualize> ;
+    schema:creativeWorkStatus <https://ld.admin.ch/vocabulary/CreativeWorkStatus/Published> ;
+    cube:observationConstraint ?shape .
```
Done.

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -249,7 +249,22 @@ const mkScoresQuery = (
   );
-  return SELECT.DISTINCT`
+
+  // HOTFIX WRT Stardog v9.2.1 bug see https://control.vshn.net/tickets/sbar-1066
```
Done.

ptbrowne commented on app/rdf/query-cube-termsets.ts:
```diff
@@ -37,3 +38,49 @@ SELECT DISTINCT ?termsetIri ?termsetLabel WHERE {
+  VALUES (?iri) {(<${iri}>)}
+
+  ?iri cube:observationConstraint/sh:property ?dimension .
+  ?dimension sh:path ?dimensionIri .
+  ?dimension a cube:KeyDimension .
+  ?dimension sh:in/rdf:rest*/rdf:first ?value.
```
Done.

ptbrowne commented on app/rdf/query-cube-termsets.ts:
```diff
@@ -37,3 +38,49 @@ SELECT DISTINCT ?termsetIri ?termsetLabel WHERE {
+PREFIX sh: <http://www.w3.org/ns/shacl#>
+
+SELECT DISTINCT ?dimensionIri ?dimensionLabel ?termsetIri ?termsetLabel WHERE {
+  VALUES (?iri) {(<${iri}>)}
```
Yep. I think I just prefer this form because in query where ?iri is repeated, it is just easier when copy pasting the built query in lindas to change the ?iri at one place.

ptbrowne commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -143,9 +189,39 @@ export const DatasetDialog = ({
+    } = ev;
+    setSharedTermsets(
+      // On autofill we get a stringified value.
+      typeof value === "string" ? value.split(",") : value
```
I think this part was part of the example of MUI, not sure if it is necessary there to make the refactor.

bprusinowski commented on app/rdf/query-search.ts:
```diff
@@ -165,27 +165,28 @@ const mkScoresQuery = (
-      ?creatorIri schema:name ?creatorLabel .
-      ?themeIri schema:name ?themeLabel .
-      ?subthemeIri schema:inDefinedTermSet ?subthemeTermset ;
+    ?dimensionIri
+      visualize:hasTermset ?termsetIri ;
+      visualize:hasTimeUnit ?unitType ;
+      schema:name ?dimensionLabel .
```
💯

ptbrowne commented:
> One thing I noticed currently on TEST: we can merge the same cube to the currently selected one, which results in no joinBy dimension error. Just thought I would drop it here, maybe it could be connected to this PR? 🤔

No longer possible.

> I could remove dimension filters completely and try to add non-compatible cube (e.g. Forest fire danger), which didn't do anything

Added a text when nothing is selected, and paused the query.

> Haven't checked the design, but feels like termsets could also be yellow? 🤔

Updated this, the termset tags are gray now.

> 
    Looks like the editor filters are not aligned with interactive time range filters (might not be related to this PR), see 2009 in the editor and 2014 in interactive filter 👀


Don't know about that, I will check later.

bprusinowski commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -138,6 +146,177 @@ const inferJoinBy = (
 
 export type JoinBy = ReturnType<typeof inferJoinBy>;
 
+export const PreviewDataTable: React.FC<{
```
I think we shouldn't use `React.FC`, it's not recommended anymore even by React core team :) See e.g. https://github.com/facebook/create-react-app/pull/8177

ptbrowne replied:
Ah it's done automatically by my refactoring extension, sorry 🙈 , thanks for the link !

bprusinowski commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -503,40 +730,15 @@ const useAddDataset = () => {
-        // from the search result.
-        const componentQueryResult = await executeDataCubesComponentsQuery(
-          client,
-          {
```
👍

bprusinowski commented on app/configurator/components/add-dataset-dialog.tsx:
```diff
@@ -318,6 +341,8 @@ export const PreviewDataTable: React.FC<{
   );
 };
 
+const MotionPaper = motion(Paper);
```
👁️ Looks like it's not used?
